sdk/src/RaygunClient.ts

```diff
@@ -0,0 +1,235 @@
+  }
+
+  initialised = true;
+
+  return true;
+};
+
+//-------------------------------------------------------------------------------------------------
+// RAYGUN CLIENT SESSION LOGIC
+//-------------------------------------------------------------------------------------------------
+
+const addTag = (...tags: string[]) => {
+  tags.forEach(tag => {
+    curSession.tags.add(tag);
+  });
+  if (!Options.disableNativeCrashReporting) {
+    RaygunNativeBridge.setTags([...curSession.tags]);
+  }
+};
+
+
+const setUser = (user: User | string) => {
+  const userObj = Object.assign(
+    {firstName: '', fullName: '', email: '', isAnonymous: false},
+    typeof user === 'string' ?
+      !!user ?
+        {identifier: user}
+        : {identifier: `anonymous-${getDeviceBasedId()}`, isAnonymous: true}
+      : user
+  );
+  curSession.user = userObj;
+  if (!Options.disableNativeCrashReporting) {
+    RaygunNativeBridge.setUser(userObj);
+  }
+};
+
+
+const clearSession = () => {
+  curSession = getCleanSession();
+  if (!Options.disableNativeCrashReporting) {
+    RaygunNativeBridge.clearSession();
+  }
+
+  crashReporter.resetCrashReporter();
+};
+//-------------------------------------------------------------------------------------------------
+// CRASH REPORTING LOGIC
+//-------------------------------------------------------------------------------------------------
+/**
+ * Converts an incoming error and its stacktrace to a standard Raygun Crash Report format.
+ * @param error
+ * @param stackFrames
+ */
+const generateCrashReportPayload = (error: Error, stackFrames: StackFrame[]) => {
+  if (CrashReportingUnavailable()) return;
+  crashReporter.generateCrashReportPayload(error, stackFrames).then();
+};
+
+
+/**
+ * Create a breadcrumb in the current session.
+ * @param message
+ * @param details
+ */
+const recordBreadcrumb = (message: string, details?: BreadcrumbOption) => {
+  if (CrashReportingUnavailable()) return;
+  crashReporter.recordBreadcrumb(message, details);
+
+};
+const sendCustomError = async (error: Error, ...params: any) => {
+  if (CrashReportingUnavailable()) return;
+
+  const [customData, tags] = (params.length == 1 && Array.isArray(params[0])) ? [null, params[0]] : params;
+
+  if (customData) {
+    addCustomData(customData as CustomData);
+  }
+  if (tags && tags.length) {
+    addTag(...tags as string[]);
+  }
+
+  await crashReporter.processUnhandledError(error);
+};
+const addCustomData = (customData: CustomData) => {
+  if (CrashReportingUnavailable()) return;
+  crashReporter.addCustomData(customData);
+
+}
+const updateCustomData = (updater: (customData: CustomData) => CustomData) => {
+  if (CrashReportingUnavailable()) return;
+  crashReporter.updateCustomData(updater);
+}
+
+/**
+ * Checks whether or not the user has initialised the client AND enabled crash reporting.
+ * @constructor
+ */
+const CrashReportingUnavailable = () => {
```

Comment:
It is preferable to have this method return a boolean and leave the logging to the method that called it. The logging can then be specific to the feature the customer is attempting to use at the wrong time. This helps customer narrow down where they are doing things out of order. They could have a race condition where they are setting a breadcrumb for an initial view before initing the client and specific logs help with this.

Reply:
Boosh